src/aks-preview/azext_aks_preview/custom.py

```diff
@@ -1582,35 +1591,25 @@ def aks_upgrade(cmd,    # pylint: disable=unused-argument
                    "upgraded to {} as well. Continue?").format(instance.kubernetes_version)
             if not prompt_y_n(msg, default="n"):
                 return None
-        upgrade_all = True
-    else:
-        if not control_plane_only:
-            msg = ("Since control-plane-only argument is not specified, this will upgrade the control plane "
-                   "AND all nodepools to version {}. Continue?").format(instance.kubernetes_version)
-            if not prompt_y_n(msg, default="n"):
-                return None
-            upgrade_all = True
-        else:
-            msg = ("Since control-plane-only argument is specified, this will upgrade only the control plane to {}. "
-                   "Node pool will not change. Continue?").format(instance.kubernetes_version)
-            if not prompt_y_n(msg, default="n"):
-                return None
-
-    if upgrade_all:
         for agent_profile in instance.agent_pool_profiles:
             agent_profile.orchestrator_version = kubernetes_version
+        return sdk_no_wait(no_wait, client.create_or_update, resource_group_name, name, instance)
 
-    # null out the SP and AAD profile because otherwise validation complains
-    instance.service_principal_profile = None
-    instance.aad_profile = None
-
-    return sdk_no_wait(no_wait, client.create_or_update, resource_group_name, name, instance)
-
+    if not control_plane_only:
+        msg = ("Since control-plane-only argument is not specified, this will upgrade the control plane "
+                "AND all nodepools to version {}. Continue?").format(instance.kubernetes_version)
+    else:
+        msg = ("Since control-plane-only argument is specified, this will upgrade only the control plane to {}. "
+                "Node pool will not change. Continue?").format(instance.kubernetes_version)
+    
+    if not prompt_y_n(msg, default="n"):
+        return None
 
-def _upgrade_single_agent_pool_node_image(client, resource_group_name, cluster_name, agent_pool_profile, no_wait):
-    instance = client.get(resource_group_name, cluster_name, agent_pool_profile.name)
-    instance.node_image_version = 'latest'
-    return sdk_no_wait(no_wait, client.create_or_update, resource_group_name, cluster_name, agent_pool_profile.name, instance)
+    cp_upgrade = sdk_no_wait(no_wait, client.create_or_update, resource_group_name, name, instance)
+    if control_plane_only: 
+        return cp_upgrade
+    #upgrade node pools one at a time
+    return upgrade_agent_pools(lambda pool: pool.orchestrator_version = kubernetes_version)
```

Comment:
munge togther cf_upgrade and the pool objects? Or  do another get on cluster and return that?